pkg/shell/shell_test.go

```diff
@@ -68,3 +69,21 @@ func testSHLVL(t *testing.T, wantSHLVL string) {
 		t.Errorf("SHLVL existence not restored, %v -> %v", oldOK, newOK)
 	}
 }
+
+// This test is unusual in that its primary purpose is to verify we don't
+// deadlock if the shell produces more output than can be buffered by a pipe
+// or tty.
+func TestScript_Deadlock(t *testing.T) {
```

Comment:
* This test should be in `progtest` since it tests the fixture defined in that package.

  The use of `Script` here is really irrelevant for the purpose of testing that the pipe in the test fixture is able to swallow data without deadline.

* "`TestScript_Deadlock`" is not very informative regarding what it really tests; there are many possibilities of deadlock. Instead of adding a comment on what a test does, use a longer function name: `TestFixture_PipeAcceptsLargeInput`.

Reply:
TL;DR:

Once there is a "real" elvish test that produces a huge amount of output (more than can be buffered by the pipe or tty) this test is redundant.

Longer answer:

I'm happy to give it a better name but I can't see much point in moving this to a *progtest_test.go* module that doesn't currently exist. While this test is to validate the behavior of the test framework its primary purpose is to verify we don't deadlock in the context of an actual elvish shell instance that produces a huge amount of output.

Ideally this would be run before practically any other test but I couldn't find a way to arrange for that to happen. If that prioritization were possible then that would justify moving this into a *progtest_test.go* module. This is, obviously, meant to tell us today whether future tests might deadlock due to producing a large amount of output. If there were such a test, such as I wrote for #661 involving output to a tty, it would deadlock if this test also deadlocked. So in a very real sense this test is redundant. It exists solely to provide confidence that any future elvish test that produces a huge amount of output (such as the one I wrote for issue #661) won't deadlock.

I'd rather drop this test than go to the trouble of moving it to a test module that doesn't currently exist since its value will be short lived.

Reply:
> Once there is a "real" elvish test that produces a huge amount of output (more than can be buffered by the pipe or tty) this test is redundant.

There will always be some redundancy between unit tests and integration tests (the "real" test you refer to). This particular test is a unit test against the `progtest` package.

> I'm happy to give it a better name but I can't see much point in moving this to a progtest_test.go module that doesn't currently exist. While this test validates the behavior of the test framework its primary purpose is to verify we don't deadlock in the context of an actual elvish shell instance that produces a huge amount of output.

Many the unit tests serve double purposes - validating some local property, and ensuring that this local property is useful in a wider context. That doesn't change the fact that unit tests are unit tests and not integration tests.

> I'd rather drop this test than go to the trouble of moving it to a test module that doesn't currently exist since its value will be short lived.

This test will still be useful when integration tests exist; in general redundant unit tests and integration tests *are* useful, because unit test failures are more focused. Imagine that the IO deadlock issue is re-introduced into `progtest`, which will cause both this test (a unit test) and a "real" elvish test (an integration test) to fail. The integration test will be harder to debug since it exercises more code paths, whereas the failure in the unit test provides very focused diagnostics that it's because the `progtest` package is misbehaving.

Reply:
Fair enough. I've moved that test to a new *progtest_test.go* file.